### PR TITLE
Render globe tiles if any layer is loaded

### LINKED ISSUE #7974: Long imagery timeout starves other imagery layers
This was [reported on the forum](https://groups.google.com/d/msg/cesium-dev/USnvihDHUic/LWJ1eQqrBAAJ). I initially thought it was the same imagery request issue I fixed here https://github.com/AnalyticalGraphicsInc/cesium/pull/7914, but this is different.

If you open [this Sandcastle example](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=xVn9ctu4EX8V1M0cpYlEykk6c5VlTx0nk2rqnK+Wc5lM7LkBSVhCBAI8AJSl3vg1+iD9u0/TJ+kuSIqfjpNcPzSTWAL384fdxS64oZpsOLtjmhwTye7IGTM8S/yf3NrAi9zPMyUt5ZJpb0R+vZYEPiE17JzumP6RR2vgnpJbKgy7lvfDo2u5AbE8oUumd47IgPRcjd9YBtKcGB++VTETQFhoEDnjlHy8GbV0Nlaz1K3BksyEKBZjdSd7lg0TLLIs7nnEzcI9pKFg5ePbTEaWKzlwtgxLy/Cjmc20JHbFjV/Z5XMZs+3FbclwckwmRznTfaFHU25YR/aIOM6GigZSvuMr5B5VRHvg/BKH4xy5XpoKluPaao70x7fUrvyEbgeTwhoyJofDm5qgLI2pzV0958YO6obcAYu68w2zVzxhKrODvYfgVa+d/XbhnhwBXORwMmxhJ9Rdz748jp3j+zLsHkCFy0H7iS+YXNoVglTi9bSFV7+D7e35f6MaUXnZG5TzDqxF1FdPyUknwEHcee8+fZE4SBjy3Xf1pVl7M/ewl4qh6BQ1p1Yh6kBUy67gFDVuLVW0Bkh9q2m0rvZ36IhK26Fm2AJcM9hbHwTkTDPYNkKFgCrAnOqyZtkVta40ELalSSoY7COQmSxNlbb+XsTVipkmJ9VMehaAAbE7YlIWcSp8QubWMyRVxnAoT8QqsqIbRpJMWI7i1S3akOwFM4llLCZghpIRG5FPmbFE8DVztir4TxcqRyTMLOEWSiYzqDuha5QcrcBzieaxiGZY2QvZ6HCurzI7UhsQiKKZtFwzshQqBGxkjB4RldJfMlb4TeP4ZbkfF6kLjyoi8OO95HJJ3tLUkFOmwX9v1CTIIC5u4TiKh0doEKqNMg3I2RqWX6/sUtG4rap2JCIh0s3zaPxRqw2Poaz82uRwBmoB0e+trE3NNAhitvE3XNuMCkY1lBPJbFsRfhKaLuxOsGlbpVv1Ly9OXzWZ7odlLj/u56mO3swX5L3SIiYLqxmzzu3PeIws3ADRgmnY36933AAfhxbCpzpacqOkgF3zI5UEuTWBZsbuqQJn28+5bT+D2mCv2vt2vy9SJnORfd4WnkYulxukbW+/QiVw/zUDz8h/SvfjSCvLBTscGz/5JYqlgxhXTHDoT/xJoEwS/AYMF5YmTJL/thPGqRk7w33qGyP8W2qs2GHCBHegQEdKKB30Jc8tcL3eWihZ4AAI/ZQu+8jA0Jjbqdsk4jSRcEcKB1+BM0s5cgVGk7Mz8vIDeQ4AklfUUqRrONikW5z6vwHhHyichFSQ11gfyHxOBkJFVAw/C/cVmF+kCOTPF+NdyAgzLmI48DLB3mkx8E4NnHQmuGJbMAUipzDJWTSfe8Nvd+7d4s2CLFY0hhPpkgnObgmctpS8f3u1GH6m/rxnIbiHXn6tiy6kIKLwOICq6kuKdlGB35dqE+T1qFV+0MzcytzICyl2VQkK0Nq+iBLFsOD18fcxGKzmyACnGIUTvDealU7g8AYi1/wEn1LWG87WhYDVfLtgdv6qJnXy7Psk6T9ltjzJknO2YWJKDv/4YJI4n3yy8MkbBmm35BCQZJEBGLveSG93RnguQ3DwHPqiVdjHzOn+yecO58uLv7z+gZyffnh9+WigfFOQ5OdzovxQQTZDOTBLplz9hL8m3/m7xDy48TiKehBW05jqdR9VSjXUFptT9tjidlFTaVLqOhgQZ3XG+kT1BUYql16X8v6BZIVOudHb9uSuBGtHZctdgrjve7HJrs8vHFLZ7lIGDWGLhRwfHxNv36p5jcZfFFNJs7NfMjuY9A9ozcG9OUPdEyag6+sRXwuReU3ToO1dNUjUrh58RAKFwJ+j9gWEn2ZmVU2UHVT7o7sX2xF01OmKjohZqbsu0B2UQHo7xB/wJ3fDSQcpBRBFefiJiowNCs0T/w9NHjTlARZ8NCIYok2WLlr9Y1Yxsn/0nG5vRDyUiH+R1bvpwtmZj+sYSahk5cDXNycWtnSGd5MKqBV41fHAXF96BwkHpxXeZ7kputSG4ycsuQseMh7zRnh3RLpo6QY7H9ZCzzndGDbhWd/dwLUMApgMYldgq4szGApx4dXFW8gHBu2MNcWcRt7N85E0wtkNHeI4hKJTVikRUgyyWEUZMqFhr3P+l7t5PPAKEg9Vt7eUpqnYoS0wqJhqgh6Vcoe98zZouAixtuKQWufyGmnuDX2ThSbSPGTVtcc+B+uz+J8p1m4SrahcQkdXQBFrlY7xEqQ2FhZXgEr7VQxRkLzJheZXDtWtXTvEHgiWvliZYHzMKu4j8vRpM0y464NKgZ0LyM6N00d+Mxy2z5Ae4/lRkySE83hdW7vfx1wPAn13YG0dNzV0ilJRI3HVo0ZRFqA6iVsrz5DmPSccxhuoTBXxKL9aHvaSQzWsQmLUSNAOMnid2a7kZamr2Vw921tes7YvI+/zOD8YHcxcc3eSk/6JJ3jfg+3GwPcDy5JUAK8JwixaM+tHxpT2/L5MxNruhlAulxrbkinRy5AOXjwbkfLfxP++flKm2GjJ5ZS8SLe15VBpOA/GmsY8M/WH9y21XKaZbZQwpi02fGMqYCSakoTHkGFdjWOr0il51tBaPgqVtSqpP22rdcEOh0lds2uGsM2Z5l8Rsg/NvsCtu/N1WpEDJC8MYbB5Yygyjyj0s/Rxnc+fp9uH1Eol2WM6XOl5VMv4y9XMgjK8ZjHfEB4fXx+03tBcH5BIUGPgyW0mxIL/jV0fnMwCoG+wCUVxjy5gnyHJkWR1eHKeL/q+PwvgZ5ercBHIYdX5eTKzoYp3BJKCjkM4CVCxgoITraZly39Q5MMMNrpOB/EPLTFk1JQ8ybvf6hobmscnSDtyr3IqgtqddkkCyJQaci3xySyPZ+xMUc+KRetQbQGbhnpcZpBcmPsOJOBsymnW0Rn06LIhwsKwPCV5W7fh7nJ2Sn5X2tqt6M7eodOFsjryHXlDw17qY0JHRLkm01SkVbu6f3jlDM5bLTAZO7qKvnH65ka6lTooj2LUQF7jeQywJ1zCrwl+o1v4dgjfjGUpLvqTw9bGFGYVzan79c5VXZx6ULzX3PCuTbMwg9IjSzPyX7XUyJNmvF9vhAU0huvauwt8z1IilL+EKwLziXvpgxeD+F6ls1Hla5XBng6s/tff/zELcrUn/0O78xdgX2z3eU7esPuffXbDV42lIHBFAJ+4moAreen4Nw), you'll see no imagery loading for a good 20-30 seconds. Even after it loads, any new imagery requests are similarly delayed. This very long delay goes away if you uncheck the "Broken Layer". 

The "broken layer" is a URL (`https://demo.boundlessgeo.com/geoserver/wms`) which takes 20-30 seconds to time out with the error Chrome error:

```
Failed to load resource: net::ERR_CONNECTION_TIMED_OUT
```

CesiumJS catches this fine and returns `An error occurred in "WebMapServiceImageryProvider": Failed to obtain image tile X: 1 Y: 0 Level: 0.` I think this is a bug because:

### What I expect to happen

1. Requests for Bing imagery go out
2. Requests for broken layer go out
3. Bing imagery requests succeed, more Bing requests go out

So overall the scene continues to load even though one imagery layer takes forever to respond.

### Instead what happens

1. Requests for Bing imagery go out
2. Requests for broken layer go out
3. Bing imagery requests succeed, **but no more Bing requests go out until the broken layer's requests respond**.

So the scene is frozen and the working imagery layer(s) do not load.

For whoever is looking into this, some helpful places to start:

* [ImageryLayer.js#L790](https://github.com/AnalyticalGraphicsInc/cesium/blob/27ef4816ea9588b58561994490667414b5513516/Source/Scene/ImageryLayer.js#L790) is where the requests go out. Notice it's throttled by server, but not throttled generally. I _believe_ this means CesiumJS keeps track of how many requests are waiting on each server, so it should be able to handle this case where one slow server doesn't starve everything else.
* [RequestScheduler.js#L64-L78](https://github.com/AnalyticalGraphicsInc/cesium/blob/27ef4816ea9588b58561994490667414b5513516/Source/Core/RequestScheduler.js#L64-L78) has the default number of maximum requests in general, and per server.
* [ImageryProvider.js#L355](https://github.com/AnalyticalGraphicsInc/cesium/blob/27ef4816ea9588b58561994490667414b5513516/Source/Scene/ImageryProvider.js#L355) this is where the actual image is fetched, and the promise is created (this is the promise returneed to ImageryLayer.js above). Beyond that, you can go into Resource.js to see what happens when you fetch an image.

### PR DESCRIPTION
**Pull request by sfariaNG**
Updated globe tiles to render imagery if any available image layer has finished loading.
Fixes #7974

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thank you so much for the pull request @sfariaNG! I noticed this is your first pull request and I wanted to say welcome to the Cesium community!

The [Pull Request Guidelines](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines) is a handy reference for making sure your PR gets accepted quickly, so make sure to skim that.

* :x: Missing `CONTRIBUTORS.md` entry.
    * Please add yourself to the [contributors](https://github.com/sfariaNG/cesium/blob/imagery-timeouts/CONTRIBUTORS.md) file!
* :x: Missing CLA.
    * Please send in a [Contributor License Agreement](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla) (CLA) and comment back here to let us know to check this!
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/imagery-timeouts/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by sfariaNG** [comment]
Guidelines addressed. Not sure a new unit test is necessary for this fix since it's so small, but let me know if you think otherwise. This should be ready.

**Comment by lilleyse** [comment]
@sfariaNG just checking - did you fill out a [CLA](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla)?

**Comment by lilleyse** [comment]
> @sfariaNG just checking - did you fill out a CLA?

Never mind.

**Comment by tfili** [comment]
@kring Do you have any input here? I’m not sure if there are any implications here that I’m missing.

File: Source/Scene/GlobeSurfaceTile.js [hunk 1]
```diff
@@ -337,14 +338,16 @@ define([
...
 
             isUpsampledOnly = isUpsampledOnly && defined(tileImagery.loadingImagery) &&
                               (tileImagery.loadingImagery.state === ImageryState.FAILED || tileImagery.loadingImagery.state === ImageryState.INVALID);
         }
 
         tile.upsampledFromParent = isUpsampledOnly;
-        tile.renderable = isRenderable;
+
+        // Allow rendering if any available layers are loaded
+        tile.renderable = isRenderable && isAnyTileLoaded;
```
Comment (nmschulte-aviture) [inline Source/Scene/GlobeSurfaceTile.js]: No more need for `isRenderable` variable; this reads more cleanly as `tile.renderable &= isAnyTileLoaded` (or `tile.renderable = tile.renderable && isAnyTileLoaded`).

Comment (sfariaNG) [reply Source/Scene/GlobeSurfaceTile.js]: That's true. I would lean toward the latter option to maintain the boolean.

**Comment by mortac8** [comment]
Is there anything else that needs to be done for this?  We are currently including it manually in our Cesium builds.

**Comment by OmarShehata** [comment]
@mortac8 I think we just haven't had a reviewer dedicate time to looking into this - hopefully we can get it looked at before the next release.

**Comment by likangning93** [comment]
@sfariaNG I grabbed a copy of the branch and merged in master, for some reason it looks like apps that load terrain on startup like the "Cesium Viewer" checkbox item above (https://cesiumjs.org/Cesium/Apps/CesiumViewer/index.html at https://cesiumjs.org/Cesium) don't render the globe until the imagery layers have been changed. Can you confirm?

Interestingly enough, if I put a breakpoint anywhere in `GlobeSurfaceTile.prototype.processImagery` I can get imagery to load, so I wonder if requests are getting cancelled due to overload somehow.

Looking at the network tab in a Sandcastle with just Cesium world terrain and Natural Earth imagery, it looks like a couple requests for both imagery and terrain are still getting through. Here's what I tested with specifically:

```js
var viewer = new Cesium.Viewer('cesiumContainer', {
    imageryProvider : Cesium.createTileMapServiceImageryProvider({
        url : Cesium.buildModuleUrl('Assets/Textures/NaturalEarthII')
    }),
    terrainProvider: Cesium.createWorldTerrain()
});
```

If it helps, I also tried logging out some tile state at the end of `GlobeSurfaceTile.prototype.processImagery`:

```js
console.log(tile.x + ' ' + tile.y + ' renderable: ' + tile.renderable + ' isDoneLoading: ' + isDoneLoading + ' isAnyTileLoaded: ' + isAnyTileLoaded);
```

Here it kind of looks like `isAnyTileLoaded` doesn't become true even though `isDoneLoading` becomes true:

Screenshot_2019-09-11_21-26-18

**Comment by likangning93** [comment]
I'm seeing something similar in IE11 and Chrome 77 on Windows 10, with the difference that I don't seem to need Cesium World Terrain to achieve the effect, just the default Bing maps imagery layer.

**Comment by likangning93** [comment]
> I'm seeing something similar in IE11 and Chrome 77 on Windows 10, with the difference that I don't seem to need Cesium World Terrain to achieve the effect, just the default Bing maps imagery layer.

Wait, no, this might just be due to an expired API key. I'm still seeing the same problems with Natural Earth imagery and terrain though.

**Comment by OmarShehata** [comment]
I can confirm this fixes the original issue referenced! I can also confirm I see @likangning93 's bug. 

@mortac8 this one isn't an intermittent issue though. If you try any Sandcastle that loads terrain on startup in this branch, like this one:

http://localhost:8080/Apps/Sandcastle/index.html?src=3D%20Tiles%20Photogrammetry.html

(You'll need to add in a token or merge the latest master branch)

You'll see it does not load any tiles. So this is going to break a lot of CesiumJS applications if merged as-is.

**Comment by mortac8** [comment]
What if you back out the change and make just one modification to GlobeSurfaceTile.js on master.   This works for me:

_Line 340:_
isRenderable = isRenderable || (thisTileDoneLoading || defined(tileImagery.readyImagery));

**Comment by sfariaNG** [comment]
I will take another look.

**Comment by sfariaNG** [comment]
Turns out just because layers are in the collection doesn't mean they are being loaded so my initial isAnyTileLoaded value wasn't always accurate. I've adjusted it to account for the case where no available layers are currently being loaded.

**Comment by sfariaNG** [comment]
This is ready for another look.

**Comment by sfariaNG** [comment]
Took another look today and realized the logic could be simplified (also helped by re-reading the comment from @likangning93 ). Ran it through 128 input combinations and got the same results as the original code except where expected (i.e. one layer loaded out of group).

[CesiumTruthTableResults2.xlsx](https://github.com/AnalyticalGraphicsInc/cesium/files/3624253/CesiumTruthTableResults2.xlsx)

**Comment by likangning93** [comment]
> Works (or fails gracefully) in IE11.

Tested in another branch with `master` merged in for https://github.com/AnalyticalGraphicsInc/cesium/pull/8188

**Comment by likangning93** [comment]
@sfariaNG thanks for the updates, and for the extensive testing! And @mortac8 thanks for your input and suggestions.

This looks good on my end and works as advertised, @sfariaNG please just merge in `master` and resolve the conflict in `CHANGES.md`.

**Comment by sfariaNG** [comment]
I resolved the conflict using the UI and merged in master, though the CI seems to not like that. Let me know if I should rebase it.

**Comment by likangning93** [comment]
@sfariaNG I'm not sure what's going on with appveyor, but I think we're ok to just go ahead and merge. Thanks again!

**Comment by sfariaNG** [comment]
Thanks @likangning93 !

**Pull request merged**